Mathlib/Combinatorics/Graph/Basic.lean

```diff
@@ -0,0 +1,325 @@
+  eq_or_eq_of_isLink_of_isLink : ∀ ⦃e x y v w⦄, IsLink e x y → IsLink e v w → x = v ∨ x = w
+  /-- An edge `e` is incident to something if and only if `e` is in the edge set. -/
+  edge_mem_iff_exists_isLink : ∀ e, e ∈ edgeSet ↔ ∃ x y, IsLink e x y
+  /-- If some edge `e` is incident to `x`, then `x ∈ V`. -/
+  left_mem_of_isLink : ∀ ⦃e x y⦄, IsLink e x y → x ∈ vertexSet
+
+namespace Graph
+
+variable {G : Graph α β}
+
+/-- `V(G)` denotes the `vertexSet` of a graph `G`. -/
+scoped notation "V(" G ")" => Graph.vertexSet G
+
+/-- `E(G)` denotes the `edgeSet` of a graph `G`. -/
+scoped notation "E(" G ")" => Graph.edgeSet G
+
+/-! ### Edge-vertex-vertex incidence -/
+
+lemma IsLink.edge_mem (h : G.IsLink e x y) : e ∈ E(G) :=
+  (edge_mem_iff_exists_isLink ..).2 ⟨x, y, h⟩
+
+lemma IsLink.symm (h : G.IsLink e x y) : G.IsLink e y x :=
+  G.isLink_symm h.edge_mem h
+
+lemma IsLink.left_mem (h : G.IsLink e x y) : x ∈ V(G) :=
+  G.left_mem_of_isLink h
+
+lemma IsLink.right_mem (h : G.IsLink e x y) : y ∈ V(G) :=
+  h.symm.left_mem
+
+lemma isLink_comm : G.IsLink e x y ↔ G.IsLink e y x :=
+  ⟨IsLink.symm, IsLink.symm⟩
+
+lemma exists_isLink_of_mem_edgeSet (h : e ∈ E(G)) : ∃ x y, G.IsLink e x y :=
+  (edge_mem_iff_exists_isLink ..).1 h
+
+lemma edgeSet_eq_setOf_exists_isLink : E(G) = {e | ∃ x y, G.IsLink e x y} :=
+  Set.ext G.edge_mem_iff_exists_isLink
+
+lemma IsLink.left_eq_or_eq_of_isLink (h : G.IsLink e x y) (h' : G.IsLink e z w) : x = z ∨ x = w :=
+  G.eq_or_eq_of_isLink_of_isLink h h'
+
+lemma IsLink.left_eq_of_isLink_of_ne (h : G.IsLink e x y) (h' : G.IsLink e z w) (hzx : x ≠ z) :
+    x = w :=
+  (h.left_eq_or_eq_of_isLink h').elim (False.elim ∘ hzx) id
+
+lemma IsLink.right_eq_of_isLink (h : G.IsLink e x y) (h' : G.IsLink e x z) : y = z := by
+  obtain rfl | rfl := h.symm.left_eq_or_eq_of_isLink h'.symm
+  · rfl
+  obtain rfl | rfl := h'.symm.left_eq_or_eq_of_isLink h.symm <;> rfl
+
+lemma IsLink.eq_and_eq_or_eq_and_eq_of_isLink {x' y' : α} (h : G.IsLink e x y)
+    (h' : G.IsLink e x' y') : (x = x' ∧ y = y') ∨ (x = y' ∧ y = x') := by
+  obtain rfl | rfl := h.left_eq_or_eq_of_isLink h'
+  · obtain rfl | rfl := h.symm.left_eq_or_eq_of_isLink h'
+    · obtain rfl | rfl := h'.symm.left_eq_or_eq_of_isLink h <;> simp
+    simp
+  obtain rfl | rfl := h.symm.left_eq_or_eq_of_isLink h'
+  · simp
+  obtain rfl | rfl := h'.left_eq_or_eq_of_isLink h <;> simp
+
+lemma IsLink.isLink_iff (h : G.IsLink e x y) {x' y' : α} :
+    G.IsLink e x' y' ↔ (x = x' ∧ y = y') ∨ (x = y' ∧ y = x') := by
+  refine ⟨h.eq_and_eq_or_eq_and_eq_of_isLink, ?_⟩
+  rintro (⟨rfl, rfl⟩ | ⟨rfl,rfl⟩)
+  · assumption
+  exact h.symm
+
+lemma IsLink.isLink_iff_sym2_eq (h : G.IsLink e x y) {x' y' : α} :
+    G.IsLink e x' y' ↔ s(x,y) = s(x',y') := by
+  rw [h.isLink_iff, Sym2.eq_iff]
+
+/-! ### Edge-vertex incidence -/
+
+/-- The unary incidence predicate of `G`. `G.Inc e x` means that `x` is one of the ends of `e`. -/
+def Inc (G : Graph α β) (e : β) (x : α) : Prop := ∃ y, G.IsLink e x y
```

Comment:
To make the naming below super clear, you could write
```suggestion
def Inc (G : Graph α β) (edge : β) (vertex : α) : Prop := ∃ y, G.IsLink edge vertex y
```
(if @kmill agrees)

Reply:
This feels a bit like overkill to me, but I'm happy to change it if outvoted.

Reply:
It looks to me like
```suggestion
def Inc (G : Graph V E) (e : E) (x : V) : Prop := ∃ y, G.IsLink e x y
```
would make it super clear too

Reply:
What about using `v` and `w` (or `v_1` and `v_2`) for vertices, instead of `x`?

Reply:
This doesn't make a difference to my own mental notation model (vertices at the end of the alphabet, edges near the beginning). I think I avoid `v` slightly because it resembles `∨` a bit too closely. But I agree that changing it here makes sense.

Reply:
Actually, in starting to make the change, the scope will naturally creep to using `G.IsLink e v w` everywhere. Do `x` and `y` seem suboptimal as vertex identifiers to you?

Reply:
I think they're fine as vertex identifiers, but I agree with Eric's point at the start that when reading `Inc` it takes a second to figure out what's going on. Maybe changing the docstring is enough though:
```
`G.Inc e x` means that the vertex `x` is one of the ends of the edge `e`.
```

Reply:
> I'm avoiding using `V` and `E` for the types altogether, because they are so close to the `V(G)` and `E(G)` notation for the vertex and edge sets, and it is important that these are different.

What about `𝒱`/`ℰ ` or some analogously curly letter?

Reply:
I'm hesitant (and not just because the curly E is hard to read). It's a small barrier to typing the symbols, and I think that anyone engaging with this stuff will internalize `Graph α β` quickly.

If I change this, I should also change `Matroid α` to `Matroid ℰ`.

Reply:
> If I change this, I should also change `Matroid α` to `Matroid ℰ`.

`Matroid α` doesn't suffer the same issue as `Graph α β` though: There's only one type, therefore it's clear what its role is.

It's generally a bad idea to give similar names to types playing different roles

Reply:
I think `x` is a fine for vertices. We use both `x,y,z` and `u,v,w` in simple graphs.

Having a good doc string here is important. How about "`G.Inc e x` means that the edge `e` is incident to the vertex `x`, which is to say that `x` is one (or both) of the endpoints of `e`"?

Reply:
Re letters to use: I don't mind alpha and beta, though of course they could be a little less obscure.

Sticking with greek, one could use `κ` for vertices (κορυφή in ancient Greek) and `γ` for edges (it's the g in poly*g*on, for line segment). I don't know how serious I am with this suggestion.

Reply:
Earlier I was using `α` and `ε`, where the `ε` suggests 'edge' for reasons less dependent on a classics degree. But Bhavik pointed out that this letter is seldom used for types. 

In response to Yael's comment, I would say that `α` and `β` do actually play similar roles, in the same sort of way as the rows/columns of a matrix. Having similar names for the types doesn't feel objectionable to me, provided conventions like `x y z u v w : α` and `e f g : β` are broadly followed.

Reply:
How about we stick with alpha and beta for now and revisit this on a future refactor?

Reply:
SGTM. I've just pushed to make docstrings more explicit throughout the file as per your suggestion.

Reply:
Sorry, I am still very much against using interchangeable letters for no-interchangeable types. Could we at least settle on non-interchangeable letters? eg `Ω α` instead of `α β`? (with the reasoning that `Ω` is the "world")

Reply:
> I'm avoiding using `V` and `E` for the types altogether, because they are so close to the `V(G)` and `E(G)` notation for the vertex and edge sets, and it is important that these are different.

I also don't *really* buy that argument: `V(G)`/`E(G)` are quite visually different to `V`/`E`. In fact, I am rather confident that if you don't name them `V`/`E` now then in six months someone (possibly me, after I have forgotten the current discussion) will whip up a quick PR renaming them to that, because it's so obviously the correct names.

Reply:
Could we please review that in six months then? If the current variable names are fine enough to last that long, I don't see any reason to hold up this PR.

Reply:
> Sorry, I am still very much against using interchangeable letters for non-interchangeable types.

I don't think these are non-interchangeable; as Peter mentions above these are just like the rows and columns of a matrix.

> I also don't really buy that argument: `V(G)`/`E(G)` are quite visually different to `V`/`E`.

I disagree with this too; to me the most worst option is for `V` to be a type and `V(G)` to be notation for a set; this feels like a very easy recipe for confusion.

Reply:
> I don't think these are non-interchangeable; as Peter mentions above these are just like the rows and columns of a matrix.

Is there an obvious map `Graph α β → Graph β α`?

Reply:
That's a high bar to set. For me, `α` is the type of 'primary things' and `β` is the 'secondary things' (or, if you like, unary things and binary things) represented by the first two letters of the Greek alphabet.

They are related like different types of radiation!